### DISCUSSION TITLE
/api/ds/query (QueryMetricsV2) should return more informative status code on query failure

### DISCUSSION THREAD

**Discussion by sd2k** [grafana/grafana#40444] [Plugins]
**What happened**:

The /api/ds/query endpoint has some decent error handling and useful error status code up until [the point where the actual queries are run](https://github.com/grafana/grafana/blob/ad757b48e7d5bb59d7a5c9ed2f9d85c2ccdf5a7e/pkg/api/metrics.go#L85-L88). If a failure occurs here, the endpoint always returns a 500 with the message `Metric request error` (the text of the actual error [isn't included in production](https://github.com/grafana/grafana/blob/78ebac0116c0aebd99d261c3f138819dc957bece/pkg/api/response/response.go#L192-L196)).

For Grafana ML we use this endpoint to fetch data from a customer's Grafana instance, but we're noticing a lot of 'Metric request error's coming back from Grafana and we're unable to make an informed decision on how to handle it without a more meaningful status or message.

**What you expected to happen**:

Ideally the error returned by `hs.DataService.HandleRequest` would be handled in an analogous way to how the `hs.handleGetDataSourceError` method works (by checking the type of the error and dispatching with a different response code & message accordingly). It might not be possible to check all errors, but dealing with some common cases would be great, before falling back to the current state.

**How to reproduce it (as minimally and precisely as possible)**:

```bash
# Start up Grafana
docker run --rm -d --name=grafana-ds-test -p 3000:3000 grafana/grafana
# Add a broken datasource
# In this case it's broken due to a bad gateway (localhost:9000 won't work inside the container)
curl -HContent-Type:application/json admin:admin@localhost:3000/api/datasources -d '{"name": "broken prometheus", "type": "prometheus", "url": "http://localhost:9000", "access": "proxy", "basicAuth": false}'
# Make a request using the /ds/query endpoint
curl -HContent-Type:application/json admin:admin@localhost:3000/api/ds/query -d '{"from": "1633949674000", "to": "1633949675000", "queries": [{"expr": "time()", "refId": "A", "datasourceId": 1, "intervalMs": 500}]}'
{"message":"Metric request error"}
```

The actual error message is available in the Grafana logs:

```
t=2021-10-11T10:57:04+0000 lvl=eror msg="Metric request error" logger=context userId=1 orgId=1 uname=admin error="Post \"http://localhost:9000/api/v1/query_range\": dial tcp 127.0.0.1:9000: connect: connection refused" remote_addr=172.17.0.1
```

Another example would be adding a datasource with some invalid auth, where the same error occurs ('Metric request error') but the logs say:
```
t=2021-10-11T11:07:21+0000 lvl=eror msg="Metric request error" logger=context userId=1 orgId=1 uname=admin error="client_error: client error: 401" remote_addr=192.168.80.1
```

**Anything else we need to know?**:

**Environment**:
- Grafana version: 8.0.0 - 8.1.7
- Data source type & version: We've experienced this with Prometheus or Graphite - but anything that can return an error from a datasource query, which is probably any of them
- OS Grafana is installed on: Linux
- User OS & Browser: N/A
- Grafana plugins: Grafana ML
- Others:

**Comment by marefr** [comment]
We don't want to expose connection information/errors to end users due to security concerns. I understand this is a problem with the use case integrating from outside Grafana. But due to the security concerns, would require each data source to implement better error handling if we add some improved sophisticated way for returning better error messages. 

#39290 is somewhat related, but not really spot on for your use case.

**Comment by marefr** [comment]
Converted this to a discussion trying to get additional feedback and better understanding of the requirements/use cases. Going to try invite  some more people.

**Comment by marefr** [comment]
To give some background of the current/historic ideas around how/why it works today. If a data source handles a query request (could be multiple queries):
- If no data frame response of any of the queries could be returned there's probably an error with the data source configuration/connection, hence why /api/ds/query returns a 500 internal server error. 
- If any of the queries results in an error there could still be queries that succeeds, hence why /api/ds/query returns a 400 bad request ([code](https://github.com/grafana/grafana/blob/ad757b48e7d5bb59d7a5c9ed2f9d85c2ccdf5a7e/pkg/api/metrics.go#L104))

**Comment by marefr** [comment]
> we're unable to make an informed decision on how to handle it without a more meaningful status or message

Can you provide examples/ideas of some of these informed decisions?

> dealing with some common cases would be great, before falling back to the current state

What do you imagine are some common cases?

**Reply by sd2k** [reply]
Sure.

> Can you provide examples/ideas of some of these informed decisions?

1. Should we retry the request? If the error was transient we may want to do this, but if it's an authentication issue, or if the datasource no longer exists, we shouldn't bother.
2. What should we deliver back to the user? We'd like to be able to say 'this error is due to a misconfiguration in your Grafana datasource, please update it', or 'the datasource was not responding, please try again later', or 'there was an issue with your query, please correct it' (this case may already be handled, actually).

I think that would actually cover some of the most common cases actually - the equivalent of having HTTP 400, 401, 403, 404, 502 and 503 error codes, falling back to 500 if neither applied.

**Reply by aocenas** [reply]
As Marcus mentioned there is this draft PR https://github.com/grafana/grafana/pull/39290 which could also cover this. The main point for it is to pass enough data to front end to know whether we can retry and if that fails to show a better message.

As mentioned we probably want to mask errors by default because we have no idea what the error is and I assume most could be just programatic errors which we should not pass on to the frontend. The draft PR gives at least ability to data sources to opt in and override this by basically saying, this error is actually safe to pass to user.

But also this only covers when it makes sense to return single error. When we do multiplexing (running multiple queries per single request) again as Marcus said there the per response error at which point not sure what makes sense cause you cannot have separate status codes per response but having custom codes like here https://github.com/grafana/grafana/pull/39290/files#diff-413beb190a3ae987b61cd69a9b7d7f71d85479a39a179cd1c1e54d74126dce56R2 could help.

**Reply by marefr** [reply]
> Should we retry the request? If the error was transient we may want to do this

This is similar to #39290. The conclusion in the discussion there was that it doesn't make sense for Grafana to return a 429 Too many requests since it's not Grafana that decided the user has done too many requests to Grafana, it's the data source responded too many requests. I think this distinction is super important, because Grafana might track metrics of it's HTTP API of requests/responses and might want to return 429 Too many requests when the user has actually done too many requests to the Grafana HTTP API. 

I'm thinking the same about 400, 401, 403, 404, 502 and 503 error codes.

>  if the datasource no longer exists 

I was under the impression you would get a 404, but looking at the code it's a 400:
https://github.com/grafana/grafana/blob/1b7ca0413ab7954b692b25a99fe5e2d181db3349/pkg/api/metrics.go#L159-L168

> What should we deliver back to the user?

My suggestion would be [422 Unprocessable Entity](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422) if there's an error returned from a data source plugin - not referring to access denied to data source or data source not exists. In addition with possibility for a plugin to provide more contextual information like a code/key categorizing the error and possibly additional information like message and additional meta data to allow you to make an informed decision. Thinking something like this and you would base informed decision on the `code` returned:

**Example request:**
```http
POST /api/ds/query HTTP/1.1
Accept: application/json
Content-Type: application/json
Authorization: Bearer eyJrIjoiT0tTcG1pUlY2RnVKZTFVaDFsNFZXdE9ZWmNrMkZYbk

{ ... }
```

**Example response:**
```http
HTTP/1.1 422 Unprocessable Entity 
Content-Type: application/json; charset=UTF-8
Content-Length: 78

{
  "code": "SERVICE_UNAVAILABLE",
  "message": "Something bad happened due to A, B, C",
  "payload": {
    "http": { "response_code": 503 }
  }
}
```

WDYT?

**Reply by sd2k** [reply]
> But also this only covers when it makes sense to return single error. When we do multiplexing (running multiple queries per single request) again as Marcus said there the per response error at which point not sure what makes sense cause you cannot have separate status codes per response but having custom codes like here https://github.com/grafana/grafana/pull/39290/files#diff-413beb190a3ae987b61cd69a9b7d7f71d85479a39a179cd1c1e54d74126dce56R2 could help.

This makes sense - a custom code per response would be ideal. Following your suggestion @marefr , perhaps something like...

```json
HTTP/1.1 422 Unprocessable Entity 
Content-Type: application/json; charset=UTF-8
Content-Length: ...

{
  "results": {
    "A": {
      "status": "success",
      "schema": ...,
      "data": ...
    },
    "B": {
      "status": "error",
      "code": "unauthorized",  // one of a few common options, or 'unknown'?
      "payload": {
        "http": { "response_code": 401 }
      }
    }
  }
}
```

Presumably the `payload` is some arbitrary object that the datasource _could_ provide here. This also proposes some standard common `code`s be included in each query's response, which would make it much easier to handle errors without worrying about the actual underlying datasource - something we need in Grafana ML since we want to support lots of datasources.

**Reply by marefr** [reply]
Sorry for late response. Trying to understand if we can find a common way to handle all of these different data source error scenarios. The tricky part is that it needs to be implemented by each plugin/developer and we want to try enforce things/guides developers through the plugin SDK somehow.

We already have this concept of notices that you can add to a data frame ([code](https://github.com/grafana/grafana-plugin-sdk-go/blob/f4ad9ec36b3b54bea8b478858b8d92eca1e0788e/data/frame_meta.go#L28-L30)). 
> Notices provide additional information about the data in the Frame that Grafana can display to the user in the user interface

What you're suggesting is the ability to make it possible to include more contextual data around an error per RefID's, right? Your use case is from the API perspective and is different than using notices I believe. In addition, notices are per frame and you suggest per RefID.

As discussed we have different scenarios of how a data source might handle queries:
- A query containing multiple queries (refID's) is making one external request to process all of the queries, e.g. Graphite data sources.
- A query containing multiple queries (refID's) is making one external request per refID to process each of the queries, e.g. Prometheus

These requests can fail in multiple ways, like connection, authentication problems for example and today depending on how a certain data source handles multiple queries you might either get a "Metric request error" or a response by RefID's that might include an error message.

Correct me if I'm wrong, what you're ultimately suggesting is if there's any error happening when processing a data source request it should always be wrapped with refID's. In the example with Graphite data sources failing to query would result in something like this:
```json
HTTP/1.1 422 Unprocessable Entity 
Content-Type: application/json; charset=UTF-8
Content-Length: ...

{
  "results": {
    "A": {
      "status": "error",
      "code": "unauthorized",  // one of a few common options, or 'unknown'?
      "payload": {
        "http": { "response_code": 401 }
      }
    },
    "B": {
      "status": "error",
      "code": "unauthorized",  // one of a few common options, or 'unknown'?
      "payload": {
        "http": { "response_code": 401 }
      }
    }
  }
}
```

Example with a single query that's failing in some way:
```json
HTTP/1.1 422 Unprocessable Entity 
Content-Type: application/json; charset=UTF-8
Content-Length: ...

{
  "results": {
    "A": {
      "status": "error",
      "code": "unauthorized",  // one of a few common options, or 'unknown'?
      "payload": {
        "http": { "response_code": 401 }
      }
    }
  }
}
```

From a consumer perspective would make it easier if the response always looked the same. However, if there's a Grafana error happening before data source handling the query it needs to be wrapped with every RefID. 

It does look a bit weird that even it's a common error for the whole request you would get the error for each RefID, but maybe it's the way forward to get some standardization of things? 

The `status` property might not be needed if we only have success/error. 

A bit unrelated to the HTTP API. Since we use protobuf/gRPC for backend plugins we might be able to use the [existing error model from Google](https://cloud.google.com/apis/design/errors). This doesn't restrict us for using a different JSON response payload for /api/ds/query though. There's some error codes in there that might help us at least.

**Comment by aocenas** [comment]
@marefr thinking about this a bit more also in context of my PR I think it would make sense to just consider 2 types of errors:
```
// This is a json response
{
  "results": {
     "A": {
        data: //frames
        // This will be filled from data source response either passing down the response error or after some processing
        error: {}
     }
  }
  // This would be filled if there was something wrong in grafana and match the status code
  "error": "error"
}
```
Basically if we return the `response.error` it will mean something inside Grafana went wrong. Otherwise we would return 200 but in the `response.results['A'].error` there can be error from the data source which we don't have control over. We could still have some partial data and as there may be multiple actual queries we could return some queries successfully and some with errors. This can be shown to the user or the data source frontend code can decide to do something with that.

I think this makes sense to separate downstream errors from our errors and also probably make our metrics clearer. Also this is basically what we already have in the:
```
type DataResponse struct {
	// The data returned from the Query. Each Frame repeats the RefID.
	Frames data.Frames

	// Error is a property to be set if the the corresponding DataQuery has an error.
	Error error
}

```

only issue is that the Error property is then sent as a single string and it would be nice to have more rich metadata about the error which could be data source specific.

**Reply by marefr** [reply]
> I think this makes sense to separate downstream errors from our errors and also probably make our metrics clearer. 

Yes, I also think this makes sense :+1: 

> Basically if we return the `response.error` it will mean something inside Grafana went wrong.

This would be similar to today if I understand you correctly, returning 500 and a Grafana error response?
```json
{
  "message": "Metric request error",
  "error": "Details" // only included when app_mode = development
}
```

I guess you also mean that getting an error back from a plugin will also result in a 500 error, like today?
https://github.com/grafana/grafana/blob/69c5370e94ed34f807ab650114450ad72a1834f9/pkg/api/metrics.go#L89-L92

> only issue is that the Error property is then sent as a single string and it would be nice to have more rich metadata about the error which could be data source specific.

Are you referring only to the Error property on DataResponse or when a Grafana error happens as well?

**Reply by aocenas** [reply]
> This would be similar to today if I understand you correctly, returning 500 and a Grafana error response?

Yes basically I changed my mind and I think the current approach makes sense :D. 

> I guess you also mean that getting an error back from a plugin will also result in a 500 error, like today?

Yes it would basically mean it's an unhandled error, probably a programatic error that plugin developer did not expect. 

> Are you referring only to the Error property on DataResponse or when a Grafana error happens as well?

I mean the Error property on DataResponse. Not sure if there is an usecase for Grafana errors.

Basically this means that if you want some reasonable message because something happened in downstream service, you have to handle it, wrap it inside DataResponse, and then it will be shown to the user as it is, or front end part of the data source can inspect it and decide what to do with it (for this we would need more than just single string though) and can do that per query.

There are few issues with this currently.
- As already mention the DataResponse.Error is only single string which isn't great for inspection.
- If there is DataResponse.Error, we change the response to 400 and then we throw exception on the front end in the same way as if it was 500. I think it would be easier to handle it on the front end if this was still 200 and plugin needed to inspect the reponse if there are errors. This is mainly weird if you have multiple queries and some were actually successful. I feel like the handling then is bit more confusing with current implementation.
- We have some types on the front end where DataResponse is converted to other type where we only have a single error property, effectively having to aggregate possible multiple errors into single one which also isn't ideal.

Not sure how feasible it is to address those though, seems like changing these things would break lots of things right now.

**Comment by dns2utf8** [comment]
Are there any docs on `QueryMetricsV2`? How should requests look? Is the API stable?

**Reply by marefr** [reply]
Sorry for slow response. When Grafana 8.5 is released the latest documentation will include this endpoint, for now you can check out the "next" documentation https://grafana.com/docs/grafana/next/http_api/data_source/#query-a-data-source

**Comment by bergquist** [comment]
I think it would be intersting to explore the idea of using http status 207 as a response as this means we can bundle multiple http status in the body instead. https://httpstatuses.com/207

**Reply by marefr** [reply]
Yes, seems like we have an agreement on this part :+1:

**Reply by marefr** [reply]
Created an issue based on our investigations of this first step https://github.com/grafana/grafana/issues/45870

**Comment by wbrowne** [comment]
Adding these issues as they don't seem to automatically get referenced and are useful for the discussion

https://github.com/grafana/grafana/issues/46337 https://github.com/grafana/grafana/issues/46338

**Reply by taleena** [reply]
Same thing happened with https://github.com/grafana/grafana/issues/48385, linking for reference.

**Comment by jcolladokuri** [comment]
I am also able to reproduce this behavior for the azure monitor data source.

The response returned by the API is a 429:
![MicrosoftTeams-image](https://user-images.githubusercontent.com/20448042/166503733-75f279f3-31c0-407b-8c86-4e834468649a.png)
 
But it gets converted to a 400:
![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/20448042/166503881-b12b9c6e-7f89-4916-88bf-d4b7e5a3a2c8.png)

 The UI does show the 429 error, but this is transient and disappears, so the user may not be aware of the throttling issue:
![MicrosoftTeams-image (2)](https://user-images.githubusercontent.com/20448042/166503980-be87f1b2-c183-40a7-a1fc-a256939838b9.png)

**Comment by taleena** [comment]
@marefr looks like the issue (https://github.com/grafana/grafana/issues/45870) fixed this for some, not all. I think we already have another issue in for the 400s, https://github.com/grafana/grafana/issues/48384, but this is 429 specifically and likely needs to account for a few of the other issues mentioned a couple comments above.

Here's another that mentions 400: https://github.com/grafana/grafana/issues/48838

**Reply by marefr** [reply]
@taleena #45870 was the first part, next part is to support plugins to return more error context/details to support #48384 and similar problems. However with this support, each datasource still needs to implement custom error handling to return a 429 as a sub-status code for a failed query - Grafana will still return HTTP response status code 207 for the query request.

**Reply by taleena** [reply]
For clarity's sake, I believe the highest priority for a fix once available is the Prom DS, since that is where we have had multiple similar reports: https://github.com/grafana/grafana/issues/48384 linked above, that also links out to https://github.com/grafana/backend-enterprise/issues/3187.

**Comment by gelicia** [comment]
Hello! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

This feature seems to be partially resolved, so I am going to close this broader discussion. If there is an aspect of this feature that was not addressed, please create a new issue detailing what you'd like to see done, and it will be triaged and prioritized along with the team's other work.
